File: languages/kotlin/k-nucleotide.kt
fun gen(length: Int): String {
    val s = CharArray(length)
    var seed = 42L
    for (i in 0 until length) {
        seed = (seed * 3877 + 29573) % 139968
        s[i] = if (seed < 42000) 'A' else if (seed < 70000) 'C' else if (seed < 98000) 'G' else 'T'
    }
    return String(s)
}

fun kNucleotide(length: Int): Long {
    val k = 8
    val p = 1000000007L
    val s = gen(length)

    val map = HashMap<String, Int>()
    var i = 0
    while (i + k <= length) {
        val kmer = s.substring(i, i + k)
        map[kmer] = (map[kmer] ?: 0) + 1
        i++
    }

    var acc = 0L
    for ((kmer, count) in map) {
        var e = 0L
        for (ch in kmer) {
            val code = when (ch) {
                'A' -> 0
                'C' -> 1
                'G' -> 2
                else -> 3
            }
            e = e * 4 + code
        }
        acc = (acc + e * count) % p
    }
    return acc
}

fun main(args: Array<String>) {
    val n = if (args.isNotEmpty()) args[0].toInt() else 100000
    println(kNucleotide(n))
    println("k-nucleotide($n)")
}

File: languages/scala/k-nucleotide.scala
import scala.collection.mutable

object KNucleotide {
  final val K = 8
  final val P = 1000000007L
  final val IM = 139968L
  final val IA = 3877L
  final val IC = 29573L

  // Deterministic DNA sequence via an integer LCG (no floating point).
  def gen(l: Int): Array[Char] = {
    val s = new Array[Char](l)
    var seed = 42L
    var i = 0
    while (i < l) {
      seed = (seed * IA + IC) % IM
      s(i) = if (seed < 42000) 'A' else if (seed < 70000) 'C' else if (seed < 98000) 'G' else 'T'
      i += 1
    }
    s
  }

  def code(c: Char): Long = c match {
    case 'A' => 0L
    case 'C' => 1L
    case 'G' => 2L
    case _   => 3L
  }

  def kNucleotide(l: Int): Long = {
    val s = gen(l)

    // Count every K-mer in the idiomatic built-in hash map, keyed by the
    // K-character substring (string) - no direct-addressing shortcut.
    val map = mutable.HashMap.empty[String, Long]
    var i = 0
    while (i + K <= l) {
      val kmer = new String(s, i, K)
      map(kmer) = map.getOrElse(kmer, 0L) + 1L
      i += 1
    }

    // Order-independent checksum: sum of encode(kmer)*count mod P.
    // e is up to 65535 and count is up to L, so e*count overflows 32-bit -
    // everything here is Long (64-bit).
    var acc = 0L
    for ((kmer, count) <- map) {
      var e = 0L
      var j = 0
      while (j < K) {
        e = e * 4L + code(kmer.charAt(j))
        j += 1
      }
      acc = (acc + e * count) % P
    }
    acc
  }

  def main(args: Array[String]): Unit = {
    val l = if (args.nonEmpty) args(0).toInt else 100000
    println(kNucleotide(l))
    println(s"k-nucleotide($l)")
  }
}
